cpp/src/arrow/io/interfaces.h

```diff
@@ -49,6 +49,26 @@ struct ReadRange {
   }
 };
 
+struct StreamMetadata {
+  struct KeyValue {
+    std::string key;
+    std::string value;
+
+    KeyValue() = default;
+    KeyValue(std::string key, std::string value)
+        : key(std::move(key)), value(std::move(value)) {}
+
+    friend bool operator==(const KeyValue& left, const KeyValue& right) {
+      return (left.key == right.key && left.value == right.value);
+    }
+    friend bool operator!=(const KeyValue& left, const KeyValue& right) {
+      return !(left == right);
+    }
+  };
+
+  std::vector<KeyValue> items;
+};
```

Comment:
I'm not sure about this API. Do we want to simplify this to `using StreamMetadata = std::vector<...>`? Do we want to add a `shared_ptr` somewhere to minimize copies (but I don't think this is performance-critical)?

Reply:
Mostly, I'd like to reuse KeyValueMetadata unless there's a reason not to

Reply:
Hmm, I had not thought about this. Interesting idea, I'll try it.